internal/backend/runtime/omni/audit/log_file.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/siderolabs/gen/pair/ordered"
+
+	"github.com/siderolabs/omni/internal/pkg/pool"
+)
+
+// LogFile is a rotating log file.
+//
+//nolint:govet
+type LogFile struct {
+	dir string
+
+	mu        sync.Mutex
+	f         *os.File
+	lastWrite time.Time
+
+	pool pool.Pool[bytes.Buffer]
+}
+
+// NewLogFile creates a new rotating log file.
+func NewLogFile(dir string) *LogFile {
+	return &LogFile{
+		dir: dir,
+		pool: pool.Pool[bytes.Buffer]{
+			New: func() *bytes.Buffer {
+				return &bytes.Buffer{}
+			},
+		},
+	}
+}
+
+// Dump writes data to the log file, creating new one on demand.
+func (l *LogFile) Dump(data any) error {
+	return l.dumpAt(data, time.Time{})
+}
+
+func (l *LogFile) dumpAt(data any, at time.Time) error {
+	b := l.pool.Get()
+	defer func() { b.Reset(); l.pool.Put(b) }()
+
+	err := json.NewEncoder(b).Encode(data)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if at.IsZero() {
+		at = time.Now()
+	}
+
+	f, err := l.openFile(at)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(f, b)
+	if err != nil {
+		return err
+	}
+
+	l.lastWrite = at
+
+	return nil
+}
+
+// openFile opens a file for the given date. It returns the file is date for at matches
+// the last write date. Otherwise, it opens a new file.
+func (l *LogFile) openFile(at time.Time) (*os.File, error) {
+	if l.f != nil && ordered.MakeTriple(at.Date()).Compare(ordered.MakeTriple(l.lastWrite.Date())) <= 0 {
+		return l.f, nil
+	}
+
+	if l.f != nil {
+		err := l.f.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	logPath := filepath.Join(l.dir, at.Format("2006-01-02")) + ".jsonlog"
+
+	f, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
```

Comment:
not exactly clear how this could would recover from a failure.

I would say `.Close()` error should be ignored (as there's no recovery path, and it doesn't actually fail).

While we should zero `l.f` to make the code jump into creating a new file.

Creating new file might fail vs. closing it

Reply:
> not exactly clear how this could would recover from a failure.

I'm not sure it can without alerting the user. We can create a temporary ring buffer if we failed to create a file, but then what? Try again at the next write (and the next after that and so on) and write all stuff which we stored if we succeed?

My current idea is to write metrics/logs if the error occured.

> While we should zero l.f to make the code jump into creating a new file.

Yeah, agreed.

Reply:
if we failed to create the file, the audit log will have missing entries, but once the error condition is clear, it should continue to log audit entries